aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,84 @@
+/**
+ * @file aesd-circular-buffer.c
+ * @brief Functions and data related to a circular buffer imlementation
+ *
+ * @author Dan Walkes
+ * @date 2020-03-01
+ * @copyright Copyright (c) 2020
+ *
+ */
+
+#ifdef __KERNEL__
+#include <linux/string.h>
+#else
+#include <string.h>
+#endif
+
+#include "aesd-circular-buffer.h"
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+  if(buffer == NULL) 
+  {
+    return NULL;
+  }
+
+  uint8_t totalEntries = buffer->full ? AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED : buffer->in_offs;
```

Comment:
@qrogers6 this logic assumes `buffer->out_offs` is 0 when the buffer is not full.